### DISCUSSION TITLE
FLUENT CUSTOM PLUGIN

### DISCUSSION THREAD

**Discussion by kandula-uday** [fluent/fluentd#4045] [Q&A]
hi All, Can anyone please help me with this? I am developing a custom plugin that will take Regular expression and search the pattern in the message

config_section :rule, param_name: :rules, multi:true do

config-param : pattern_regexp,  :regexp,  default :/^$/

def configure(conf)
 super
@rp = conf['pattern_regexp']
puts @rp
end

def filter(tag, time, record)
record.each do |key, value|
if value =~ @rp 
end
record
end
end
end

conf. file
<filter>
@mask
<rule> 
pattern_regexp /[a-zA-Z0-9.]+t$/
</rule>
</filter>

but I am getting nil class.. @rp is not holding the conf value of pattern_regexp i am getting nil error

**Comment by daipom** [comment]
I think this will be helpful.

https://github.com/fluent/fluentd/blob/v1.15.3/lib/fluent/plugin/filter_grep.rb#L44-L49
https://github.com/fluent/fluentd/blob/v1.15.3/lib/fluent/plugin/filter_grep.rb#L115-L118

```ruby
@rules.each do |rule|
  p rule.pattern_regexp
end
```

**Reply by kandula-uday** [reply]
Thank You Daijiro Fukuda it worked. May I know why @rp = conf['pattern_regexp'] not holding the value of pattern_regex from conf file

**Reply by daipom** [reply]
We can not get `config_section`  values from `conf` by `hash[]`.
I don't know the reason, but we don't need it.

We can use configured parameters by instance variables or accessors.
Usually, all we have to do is defining `config_param` or `config_section` and so on.

Overriding `configure(conf)` is for some processes after parsing the config (which is done in `super`).
Usually, we don't have to use `conf` value, since it is just for overriding.

```rb
def configure(conf)
  super
  # some processes such as validating
  validate(@rules)
end
```

**Reply by kandula-uday** [reply]
so ,any instance variables to be used in the functions are defined or initialized in def configure(conf)

**Reply by daipom** [reply]
The config values which you define by `config_param` or `config_section` are initialized as instance variables and accessors in `super` of `def configure(conf)`.
